app/models/null_user.rb

```diff
@@ -5,7 +5,7 @@ def persisted?
   end
 
   def initialize(name)
-    super({ first_name: name, created_at: Time.now })
+    super({ id: -1, first_name: name, created_at: Time.now })
```

Comment:
@NickSchimek  just wanted to know. Why did you use a negative id instead of a positive one

Reply:
To prevent database errors.

Here's a little scenario: Let's say we have a fresh database. Never been used before. We want our ghost user to have an `id` of `1`. 
`Create command:`
```
[2] pry(main)> User.create id: 1, first_name: "ghost", email: "ghost@example.org", password: "password"
```
Database transactions are successful.
```
   (0.2ms)  BEGIN
  User Exists (0.6ms)  SELECT  1 AS one FROM "users" WHERE ("users"."id" != 1) AND "users"."slug" = $1 LIMIT $2  [["slug", "ghost"], ["LIMIT", 1]]
  User Exists (0.2ms)  SELECT  1 AS one FROM "users" WHERE "users"."email" = $1 AND "users"."deleted_at" IS NULL LIMIT $2  [["email", "ghost@example.org"], ["LIMIT", 1]]
  Authentication Exists (9.2ms)  SELECT  1 AS one FROM "authentications" WHERE "authentications"."user_id" = $1 LIMIT $2  [["user_id", 1], ["LIMIT", 1]]
  Authentication Exists (0.4ms)  SELECT  1 AS one FROM "authentications" WHERE "authentications"."user_id" = $1 LIMIT $2  [["user_id", 1], ["LIMIT", 1]]
  SQL (0.8ms)  INSERT INTO "users" ("id", "email", "encrypted_password", "created_at", "updated_at", "first_name", "slug") VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING "id"  [["id", 1], ["email", "ghost@example.org"], ["encrypted_password", "$2a$10$VD6der/ks6Y8WSsCF.ToT.cS44HL3NF9nn4foGitZS8qi56UtY1eG"], ["created_at", "2018-10-09 04:22:23.536777"], ["updated_at", "2018-10-09 04:22:23.536777"], ["first_name", "ghost"], ["slug", "ghost"]]
  ActsAsTaggableOn::Tagging Load (19.7ms)  SELECT "taggings".* FROM "taggings" WHERE "taggings"."taggable_id" = $1 AND "taggings"."taggable_type" = $2  [["taggable_id", 1], ["taggable_type", "User"]]
   (0.6ms)  COMMIT
  ActsAsTaggableOn::Tag Load (0.4ms)  SELECT "tags".* FROM "tags" INNER JOIN "taggings" ON "tags"."id" = "taggings"."tag_id" WHERE "taggings"."taggable_id" = $1 AND "taggings"."taggable_type" = $2 AND (taggings.context = 'skills' AND taggings.tagger_id IS NULL)  [["taggable_id", 1], ["taggable_type", "User"]]
  ActsAsTaggableOn::Tag Load (0.3ms)  SELECT "tags".* FROM "tags" INNER JOIN "taggings" ON "tags"."id" = "taggings"."tag_id" WHERE "taggings"."taggable_id" = $1 AND "taggings"."taggable_type" = $2 AND (taggings.context = 'titles' AND taggings.tagger_id IS NULL)  [["taggable_id", 1], ["taggable_type", "User"]]
```
Here is the return value from creating our ghost user:
```
=> #<User id: 1, email: "ghost@example.org", created_at: "2018-10-09 04:22:23", updated_at: "2018-10-09 04:22:23", first_name: "ghost", last_name: nil, display_email: nil, youtube_id: nil, slug: "ghost", display_profile: true, latitude: nil, longitude: nil, country_name: nil, city: nil, region: nil, youtube_user_name: nil, github_profile_url: nil, display_hire_me: nil, bio: nil, receive_mailings: false, country_code: nil, timezone_offset: nil, status_count: 0, deleted_at: nil, event_participation_count: 0, can_see_dashboard: false, skill_list: nil, title_list: nil>
```

Alright! Now that is out of the way let's have real users start signing up online.
Here's the first User: Notice I did not specify an id number, because the database increments these automatically.
```
[3] pry(main)> User.create first_name: "real_person", email: "real_person@example.org", password: "password"
```
Database transactions: Notice the Rollback and error message.
```
   (0.1ms)  BEGIN
  User Exists (0.5ms)  SELECT  1 AS one FROM "users" WHERE ("users"."id" IS NOT NULL) AND "users"."slug" = $1 LIMIT $2  [["slug", "real_person"], ["LIMIT", 1]]
  User Exists (0.2ms)  SELECT  1 AS one FROM "users" WHERE "users"."email" = $1 AND "users"."deleted_at" IS NULL LIMIT $2  [["email", "real_person@example.org"], ["LIMIT", 1]]
  SQL (6.5ms)  INSERT INTO "users" ("email", "encrypted_password", "created_at", "updated_at", "first_name", "slug") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"  [["email", "real_person@example.org"], ["encrypted_password", "$2a$10$yDOTvzvyxWpcG4x668EXneL5Loz9phWlDU.tnQzDrqGRlrUs.Qd06"], ["created_at", "2018-10-09 04:22:46.253559"], ["updated_at", "2018-10-09 04:22:46.253559"], ["first_name", "real_person"], ["slug", "real_person"]]
   (0.2ms)  ROLLBACK
ActiveRecord::RecordNotUnique: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "users_pkey"
DETAIL:  Key (id)=(1) already exists.
: INSERT INTO "users" ("email", "encrypted_password", "created_at", "updated_at", "first_name", "slug") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"
from /Users/miafro/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-5.1.6/lib/active_record/connection_adapters/postgresql_adapter.rb:616:in `async_exec'
```

The database had a duplicate key value error, hence it tried to create the first signup using id of 1, but that was already taken because we manually added our ghost user to have an id of 1. 

So I can't use a positive integer except for `0`, because even if I use a large number say 1,000,000 eventually the database will get to the row number and error out.

so `-1` is a good value to use (`0` could be used too) unless negative numbers are being used for the database rows which they shouldn't unless there is a very good reason.

Bottom line is: Do not specify an id when manually creating database rows in production, but we needed to in order to roll out the ghost user without bugs so `-1` is our only option.

Reply:
@NickSchimek thank you for explanation. Makes more sense now